Hiding the names of files

Onimishra opened the pull request:
An open-source security library should not advertise its own use to the world. If an attacker gets access to the filesystem, he would be able to see what library that was used, simply based on the names of the files stored (SPS_file and the sps_ prefix of the keys on old devices).

With this changed to be configurable by the user, and with a good suite of obfuscation and string encryption tools, it is much harder for an attacker to find out which library (if any) was used to make the encrypted preference store.

All log statements has been removed for the same reason.

Onimishra committed:
Change: Removed logging. A security library should never advertice who they are. When going though class obfuscation (and sometimes even class encryption), it is pointless if a opensource project advertise its own use within the application, making for an easy attack vector, as the entire code base is public.

Onimishra committed:
Change: Name of the store file can now be set by the application using the library. A secure open-source library should try to make it as hard as possible to know that it is being used.

Onimishra committed:
Change: The prefix of the keys in the store can now be set på the application. This removes all references in the files that an attacker could see, that Secure Preference Store is in use within the application. The next steps for an attacker would be decompiling of the application, but all strings can be encrypted within and all classes can be obscured beyond recognition, making several extra layers, before the attacker knows what library is in use.

Onimishra committed:
Change: Readme now contains a line about hiding the use of Secure Preference Store.

iamMehedi commented:
Making the file name and key prefixes user configurable is a good addition. But I don't think that'll improve the security much. If an attacker has access to the file, he also knows which app has created it and he can easily find out which libraries that particular app is using.

iamMehedi commented on securedpreferencestore/src/main/java/devliving/online/securedpreferencestore/EncryptionManager.java:
```diff
@@ -107,8 +114,15 @@
 
     SecuredPreferenceStore.KeyStoreRecoveryNotifier mRecoveryHandler;
 
-    public EncryptionManager(Context context, SharedPreferences prefStore, SecuredPreferenceStore.KeyStoreRecoveryNotifier recoveryHandler)
```
we need to keep the older constructor in case someone wants to keep using the default.

Onimishra replied:
Is it intended for end-users to use the EncryptionManager? I see that it is not PackagePrivate, thus indicating that end-users may get a hold of it.
If I can provide a reliable upgrade mechanism, which would also address your compatibility point, do you then believe it is still necessary to keep the old constructor?

iamMehedi replied:
Yes, this is intended for the end user specially in cases where the user is not using the secured preference but wants to use the library to encrypt/decrypt data. 

Yes, it's not mandatory to keep the old constructor if we have a smooth non-confusing API to upgrade to.

iamMehedi commented on securedpreferencestore/src/main/java/devliving/online/securedpreferencestore/EncryptionManager.java:
```diff
@@ -107,8 +114,15 @@
     SecuredPreferenceStore.KeyStoreRecoveryNotifier mRecoveryHandler;
 
-    public EncryptionManager(Context context, SharedPreferences prefStore, SecuredPreferenceStore.KeyStoreRecoveryNotifier recoveryHandler)
+    public EncryptionManager(Context context, SharedPreferences prefStore, String keyAliasPrefix, SecuredPreferenceStore.KeyStoreRecoveryNotifier recoveryHandler)
```
I think key prefix should be marked with `@NonNull` here. That way user will have to distinguishably choose whether they want to provide one (so use this constructor) or not  (use the older constructor with default prefix). It'll prevent developers from using the wrong constructor without explicitly knowing.

Onimishra replied:
There are specific checks for null in the code, setting a default filename (same as in the current version). This is to support the same init methods that SecurePreferenceStore class contains.

iamMehedi replied:
In my opinion we can drop the Old constructor and make this the only constructor but in that case we should mark the `String keyAliasPrefix` as `@Nullable` thus indicating that a `null` value is ok.

iamMehedi commented on securedpreferencestore/src/main/java/devliving/online/securedpreferencestore/SecuredPreferenceStore.java:
```diff
@@ -75,16 +73,29 @@ synchronized public static SecuredPreferenceStore getSharedInstance() {
                              RecoveryHandler recoveryHandler ) throws IOException, CertificateException, NoSuchAlgorithmException, KeyStoreException, UnrecoverableEntryException, InvalidAlgorithmParameterException, NoSuchPaddingException, InvalidKeyException, NoSuchProviderException {
 
         if(mInstance != null){
-            Log.w("SECURED-PREFERENCE", "init called when there already is a non-null instance of the class");
```
logs are important for development as currently we don't have unit tests. But we should definitely not print these in production, so may be keep these but add a check to make sure that we're printing only in DEBUG builds.

Onimishra replied:
I'll be wrapping all log and printStackTrace statements in a custom logger class. This will by default only allow logging to happen when the BuildConfig.DEBUG flag is set, with the ability for the end-user to force it on, should they find it necessary.

iamMehedi replied:
Sounds good!

iamMehedi commented on securedpreferencestore/src/main/java/devliving/online/securedpreferencestore/SecuredPreferenceStore.java:
```diff
@@ -75,16 +73,29 @@ synchronized public static SecuredPreferenceStore getSharedInstance() {
+    /**
+     * @see #init(Context, String, RecoveryHandler)
+     */
+    public static void init( Context appContext, String fileName, RecoveryHandler recoveryHandler) throws IOException, CertificateException, NoSuchAlgorithmException, InvalidKeyException, UnrecoverableEntryException, InvalidAlgorithmParameterException, NoSuchPaddingException, NoSuchProviderException, KeyStoreException {
```
for the feature to enable users to change the filename and key prefixes, we need to think about backward compatibility e.g how does these changes affect existing apps that use old versions? how do they upgrade to this version without loosing data?

javichaques commented:
@iamMehedi Can you accept this pull request and publish a new realease? I need this feature to change the preferences file name.

Thank you

iamMehedi commented:
@javichaques there are a few changes pending in this pull request. As soon as those are resolved, I can merge and publish a new version.

Onimishra commented:
Sorry about not following up on this pull request. I'll try to amend the commit and submit again.

Onimishra committed:
Add: Migrations. Created first migration from 0.5.0 to 0.6.0
Add: Small logging framework. This makes it possible to turn all logging in the library on with a single call. The default options is on for DEBUG and off for everything else.
Change: Bumped version number to 0.6.0, to distinguish then new old data format from the old.
Change: Deprecated the old init function. This will initialise a version of the store, that can read and write the old data format.
Change: Annotated the encryption manager with Nullable parameters. If null is passed in all the new parameters, or the deprecated constructor is used, the manager will be able to read and write the old format.
Change: A version number is now written into the dataset. This will help determine how to update the data, should new updates be available later.

Onimishra commented:
Okay, here we go.
I ended up creating old constructors and init methods for both SecuredPeferenceStore and EncryptionManager. Using this, will initialise everything so it can read and write data in the same manner as 0.5.0. This means, people who just update will not break their apps.
All logging has been reinstated. It is now done through a logger class, making it possible to turn it on/off from central location. It is on in DEBUG mode, and can be turned on in RELEASE with a method call.
Now, to get reliably to 0.6.0. This was a bit more tricky (as all data migration always is). I've created the beginning of a migration tool. The end user can supply Migration instances, going from what was released when their app was first released. For now, that would be 0.5.0 to 0.6.0. Doing it this way, means that each migration level can get supplied with just the parameters that the specific level needs. For 0.6.0, that would be bitshift and filename. We can not change the prefix during upgrade, as we do not now the keys - they are hashes. Instead, I write an "override" prefix to the updated store, so even though the new store is initialised with "myPrefix", if the store was an update, it will still get "sps" loaded. Clean installs will have the "mePrefix".
A version number is now also present in the store, keeping track of what level the data set is at. This is nesssesary for future migrations.

As I don't have a big fleet of hands-on devices to test this on, and it's difficult to test on AWS DeviceFarm or Firebase Test Lab due to the need to upgrade an app halfway through, this has only been tested in emulators running different versions of stock Android.

Commit:
Merge branch 'master' of github.com:iamMehedi/Secured-Preference-Store

Commit:
compatibility issues resolved

iamMehedi committed:
Update README.md
